### PR TITLE
Improve loading of encrypted attributes.

### PR DESCRIPTION
**Pull request by finalwharf**
Since attribute proxy was introduced, loading/decrypting of vault attributes was done whenever the value of an attribute is `nil`. This prevented an attribute to be set to `nil` in lazy models, because the AR thinks that the attribute hasn't changed. (Lazy attributes are not loaded until they are accessed. If they are not, they are `nil`. And when we try to set them to `nil`, Active Record doesn't pick up the "change").

### CODE REVIEW

File: spec/integration/rails_spec.rb [hunk 1]
```diff
@@ -36,14 +36,20 @@
...
       person.update_attributes!(ssn: nil)
       person.reload
 
       expect(person.ssn).to be(nil)
     end
 
+    it "allows attributes to be unset" do
+      person = Person.create!(ssn: "123-45-6789")
+      person.ssn = nil
+      expect(person.ssn).to be(nil)
```
Comment (h-lame) [inline spec/integration/rails_spec.rb]: Is the expectation that the `ssn_encrypted` value would also be `nil` at this point?  Can we make that explicit?

File: spec/integration/rails_spec.rb [hunk 2]
```diff
@@ -187,6 +192,12 @@
       expect(person.ssn).to be(nil)
     end
 
+    it "allows attributes to be unset" do
+      person = LazyPerson.create!(ssn: "123-45-6789")
+      person.ssn = nil
+      expect(person.ssn).to be(nil)
+    end
```
Comment (h-lame) [inline spec/integration/rails_spec.rb]: Same comment as for the non-lazy version of this spec, what would we expect the `ssn_encrypted` value to be now and can we put that in the test?

File: lib/vault/encrypted_model.rb [hunk 3]
```diff
@@ -233,7 +256,7 @@ def __vault_persist_attributes!
         # If there are any changes to the model, update them all at once,
         # skipping any callbacks and validation. This is okay, because we are
         # already in a transaction due to the callback.
-        self.update_columns(changes) if !changes.empty?
+        self.update_columns(changes) unless changes.empty?
```
Comment (h-lame) [inline lib/vault/encrypted_model.rb]: 👏

File: lib/vault/encrypted_model.rb [hunk 4]
```diff
@@ -220,6 +241,8 @@ def __vault_load_attribute!(attribute, options)
         # Deserialize the plaintext value, if a serializer exists
         plaintext = serializer.decode(plaintext) if serializer
 
+        __loaded_attributes << attribute
```
Comment (h-lame) [inline lib/vault/encrypted_model.rb]: I think it makes sense to name space this var like the other ones and call it `__vault_loaded_attributes` just in case because `__loaded_attributes` might be too generic and end up shadowing some other internal var used by rails or another gem.

File: lib/vault/encrypted_model.rb [hunk 5]
```diff
@@ -301,7 +324,9 @@ def reload(*)
             write_attribute(attribute, nil)
           end
 
-          self.__vault_load_attributes!
+          @__loaded_attributes = Set.new
```
Comment (h-lame) [inline lib/vault/encrypted_model.rb]: Could we `__loaded_attributes.clear` instead?  Might marginally save on memory to reuse the same set rather than creating a new one each reload.

File: lib/vault/encrypted_model.rb [hunk 6]
```diff
@@ -79,14 +79,27 @@ def vault_attribute(attribute_name, options = {})
...
+          end
+
+          read_attribute(attribute_name)
         end
 
         # Setter
         define_method("#{attribute_name}=") do |value|
+          # Prevent the attribute from loading when a value is provided before
+          # the attribute is loaded from Vault but only if the model is initialized
+          # __vault_load_attribute!(attribute_name, self.class.__vault_attributes[attribute_name])
```
Comment (h-lame) [inline lib/vault/encrypted_model.rb]: Can we drop this commented out line of code?  It doesn't seem relevant to the comment to me so it's confusing.

File: spec/integration/rails_spec.rb (hunk 1 above)
Comment (finalwharf) [reply spec/integration/rails_spec.rb]: This test is only to test that the getter does not try to reload the attribute from the encrypted column if it's set to `nil`.

File: spec/integration/rails_spec.rb (hunk 2 above)
Comment (finalwharf) [reply spec/integration/rails_spec.rb]: > This test is only to test that the getter does not try to reload the attribute from the encrypted column if it's set to nil.

**Comment by finalwharf** [comment]
@h-lame I've answered your questions and addressed your comments.
